google/cloud/spanner_v1/transaction.py

```diff
@@ -305,3 +329,59 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             self.commit()
         else:
             self.rollback()
+
+
+class ResultsChecksum:
+    """Cumulative checksum.
+
+    Used to calculate a total checksum of all the results
+    returned by operations executed within transaction.
+    Includes methods for checksums comparison.
+
+    These checksums are used while retrying an aborted
+    transaction to check if the results of a retried transaction
+    are equal to the results of the original transaction.
+    """
+
+    def __init__(self):
+        self.checksum = hashlib.sha256()
+        self.count = 0  # counter of consumed results
+
+    def __eq__(self, other):
+        """Check if checksums are equal.
+
+        Args:
+            other (ResultsChecksum):
+                Another checksum to compare with this one.
+        """
+        same_count = self.count == other.count
+        same_checksum = self.checksum.digest() == other.checksum.digest()
+        return same_count and same_checksum
+
+    def __ne__(self, other):
+        """Check if checksums aren't equal.
+
+        Args:
+            other (ResultsChecksum):
+                Another checksum to compare with this one.
+        """
+        return not self == other
+
+    def __lt__(self, other):
+        """Check if this checksum have less results than the given one.
+
+        Args:
+            other (ResultsChecksum):
+                Another checksum to compare with this one.
+        """
+        return self.count < other.count
+
+    def consume_result(self, result):
+        """Add the given result into the checksum.
+
+        Args:
+            result (Union[int, list]):
+                Streamed row or row count from an UPDATE operation.
+        """
+        self.checksum.update(pickle.dumps(result))
```

Comment:
Do you need to serialize rows before adding them to the checksum?

I still need to review the logic here, but I'm surprised to see `pickle.dumps` with user data like this. I don't see any security problem since you're not _un_-pickling results, but since you don't care about reconstructing the result rows pickle may be overkill.

Reply:
Yeah, `update()` requires bytes-like object to be passed